engine/src/main/java/org/terasology/rendering/logic/SkeletonRenderer.java

```diff
@@ -113,55 +113,72 @@ public void newSkeleton(OnActivatedComponent event, EntityRef entity) {
+        SkeletalMeshComponent skeletalMeshComp = entity.getComponent(SkeletalMeshComponent.class);
 
-            if (skeletalMeshComp.animation == null) {
-                continue;
-            }
+        if (skeletalMeshComp.animation == null && skeletalMeshComp.animationPool != null) {
+            skeletalMeshComp.animation = randomAnimationData(skeletalMeshComp, random);
+        }
 
-            if (skeletalMeshComp.animation.getFrameCount() <= 0) {
-                continue;
-            }
+        if (skeletalMeshComp.animation == null) {
+            return;
+        }
 
-            skeletalMeshComp.animationTime += delta * skeletalMeshComp.animationRate;
-            float framePos = skeletalMeshComp.animationTime / skeletalMeshComp.animation.getTimePerFrame();
+        if (skeletalMeshComp.animation.getFrameCount() < 1) {
+            return;
+        }
+        skeletalMeshComp.animationTime += delta * skeletalMeshComp.animationRate;
+        float animationDuration = getDurationOfAnimation(skeletalMeshComp);
+        while (skeletalMeshComp.animationTime >= animationDuration) {
+            MeshAnimation newAnimation;
+            if (!skeletalMeshComp.loop) {
+                newAnimation = null;
+            } else {
+                newAnimation = randomAnimationData(skeletalMeshComp, random);
+            }
 
-            if ((int) framePos >= skeletalMeshComp.animation.getFrameCount()) {
-                while ((int) framePos >= skeletalMeshComp.animation.getFrameCount()) {
-                    framePos -= skeletalMeshComp.animation.getFrameCount();
-                    skeletalMeshComp.animationTime -= skeletalMeshComp.animation.getTimePerFrame() * skeletalMeshComp.animation.getFrameCount();
-                }
-                if (!skeletalMeshComp.loop) {
-                    skeletalMeshComp.animation = null;
-                    entity.send(new AnimEndEvent(skeletalMeshComp.animation));
-                }
-                MeshAnimation newAnimation = randomAnimationData(skeletalMeshComp, random);
-
-                if (newAnimation == null) {
-                    MeshAnimationFrame frame = skeletalMeshComp.animation.getFrame(
-                            skeletalMeshComp.animation.getFrameCount() - 1);
-                    updateSkeleton(skeletalMeshComp, frame, frame, 1.0f);
-                    skeletalMeshComp.animation = null;
-                    entity.saveComponent(skeletalMeshComp);
-                    continue;
-                }
-                skeletalMeshComp.animation = newAnimation;
+            if (newAnimation == null) {
+                MeshAnimation finishedAnimation = skeletalMeshComp.animation;
+                skeletalMeshComp.animationTime = animationDuration;
+                /*
+                 * Set animation to null so that AnimEndEvent fires only once
+                 */
+                skeletalMeshComp.animation = null;
```

Comment:
Something is awry with the location of this line with .animation used on the next line methinks :-) Tried a quick fix simply moving it below